torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp

```diff
@@ -0,0 +1,268 @@
+
+  // For some reason I couldn't get nvjpegEncodeImage to work for grayscale
+  // images but nvjpegEncodeYUV seems to work fine. Annoyingly however,
+  // nvjpegEncodeYUV requires the source image plane pitch to be divisible by 8
+  // so we must pad the image if needed:
+  // https://github.com/NVIDIA/cuda-samples/issues/23#issuecomment-559283013
+  std::shared_ptr<torch::Tensor> padded_image_ptr =
+      std::make_shared<torch::Tensor>(src_image);
+  int padded_width = width;
+  if (channels == 1 && width % YUV_IMG_SIZE_DIVISIBILITY != 0) {
+    padded_width = ((width / YUV_IMG_SIZE_DIVISIBILITY) + 1) * YUV_IMG_SIZE_DIVISIBILITY;
+    int left_padding, right_padding;
+    left_padding = right_padding = (padded_width - width) / 2;
+    if ((padded_width - width) % 2 != 0) {
+      ++left_padding;
+    }
+    torch::nn::functional::PadFuncOptions pad_options(
+        {left_padding, right_padding});
+    auto padded_image =
+        torch::nn::functional::pad(src_image, pad_options).contiguous();
+    padded_image_ptr = std::make_shared<torch::Tensor>(padded_image);
+  }
+
+  for (int c = 0; c < channels; c++) {
+    target_image.channel[c] = (*padded_image_ptr)[c].data_ptr<uint8_t>();
+    // this is why we need contiguous tensors
+    target_image.pitch[c] = padded_width;
+  }
+  for (int c = channels; c < NVJPEG_MAX_COMPONENT; c++) {
+    target_image.channel[c] = nullptr;
+    target_image.pitch[c] = 0;
+  }
+  nvjpegStatus_t encodingState;
+  // Encode the image
+  if (channels == 1) {
+    // For some reason nvjpegEncodeImage fails for grayscale images, so we use
+    // nvjpegEncodeYUV instead
+    encodingState = nvjpegEncodeYUV(
+        nvjpeg_handle,
+        nv_enc_state,
+        nv_enc_params,
+        &target_image,
+        NVJPEG_CSS_GRAY,
+        padded_width,
+        height,
+        stream);
+  } else {
+    encodingState = nvjpegEncodeImage(
+        nvjpeg_handle,
+        nv_enc_state,
+        nv_enc_params,
+        &target_image,
+        NVJPEG_INPUT_RGB,
+        width,
+        height,
+        stream);
+  }
+  TORCH_CHECK(
+      encodingState == NVJPEG_STATUS_SUCCESS,
+      "image encoding failed: ",
+      encodingState);
+
+  // Retrieve length of the encoded image
+  size_t length;
+  nvjpegStatus_t getStreamState = nvjpegEncodeRetrieveBitstreamDevice(
+      nvjpeg_handle, nv_enc_state, NULL, &length, stream);
+  TORCH_CHECK(
+      getStreamState == NVJPEG_STATUS_SUCCESS,
+      "Failed to retrieve encoded image stream state: ",
+      getStreamState);
+
+  // Synchronize the stream to ensure that the encoded image is ready
+  cudaError_t syncState = cudaStreamSynchronize(stream);
+  TORCH_CHECK(syncState == cudaSuccess, "CUDA ERROR: ", syncState);
+
+  // Reserve buffer for the encoded image
+  torch::Tensor encoded_image = torch::empty(
+      {static_cast<long>(length)},
+      torch::TensorOptions()
+          .dtype(torch::kByte)
+          .layout(torch::kStrided)
+          .device(torch::kCUDA, 1)
```

Comment:
This fails on machines with only 1 GPU where only device `0` exists. Should we instead allocate the output tensors on teh same device as the original image (i.e. use the `device` parameter as passed to `encode_single_jpeg()`)?

Reply:
Ah yes, of course